# 4.1 Data Exploration 

Note: Credit Card Fraud dataset is downloaded from https://www.kaggle.com/datasets/mlg-ulb/creditcardfraud?resource=download

## Basic Exploration

1. Import `pandas` library
2. Import dataset

In [6]:
import pandas as pd

df = pd.read_csv('../datasets/creditcard.csv')

3. Check the number of rows and columns in the loaded data

Output: (row numbers, column numbers)

In [7]:
df.shape

(284807, 31)

4. Print information about a DataFrame including the index dtype and columns, non-null values and memory usage.

All columns say float64 next to them, indicating they are of float data type, that is, numbers such as ..., -0.4, 1.29, 0.0, 1.5, 3.5,... . The exception is the Class column. This contains integer values that indicate if a transaction is fraudulent or not. 

In [8]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 284807 entries, 0 to 284806
Data columns (total 31 columns):
 #   Column  Non-Null Count   Dtype  
---  ------  --------------   -----  
 0   Time    284807 non-null  float64
 1   V1      284807 non-null  float64
 2   V2      284807 non-null  float64
 3   V3      284807 non-null  float64
 4   V4      284807 non-null  float64
 5   V5      284807 non-null  float64
 6   V6      284807 non-null  float64
 7   V7      284807 non-null  float64
 8   V8      284807 non-null  float64
 9   V9      284807 non-null  float64
 10  V10     284807 non-null  float64
 11  V11     284807 non-null  float64
 12  V12     284807 non-null  float64
 13  V13     284807 non-null  float64
 14  V14     284807 non-null  float64
 15  V15     284807 non-null  float64
 16  V16     284807 non-null  float64
 17  V17     284807 non-null  float64
 18  V18     284807 non-null  float64
 19  V19     284807 non-null  float64
 20  V20     284807 non-null  float64
 21  V21     28

5. Print the first few rows of the DataFrame

In [9]:
df.head()

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  

[5 rows x 31 columns]

5.1 Print more...

In [10]:
df.head(10)

Time        V1        V2        V3        V4        V5        V6        V7  \
0   0.0 -1.359807 -0.072781  2.536347  1.378155 -0.338321  0.462388  0.239599   
1   0.0  1.191857  0.266151  0.166480  0.448154  0.060018 -0.082361 -0.078803   
2   1.0 -1.358354 -1.340163  1.773209  0.379780 -0.503198  1.800499  0.791461   
3   1.0 -0.966272 -0.185226  1.792993 -0.863291 -0.010309  1.247203  0.237609   
4   2.0 -1.158233  0.877737  1.548718  0.403034 -0.407193  0.095921  0.592941   
5   2.0 -0.425966  0.960523  1.141109 -0.168252  0.420987 -0.029728  0.476201   
6   4.0  1.229658  0.141004  0.045371  1.202613  0.191881  0.272708 -0.005159   
7   7.0 -0.644269  1.417964  1.074380 -0.492199  0.948934  0.428118  1.120631   
8   7.0 -0.894286  0.286157 -0.113192 -0.271526  2.669599  3.721818  0.370145   
9   9.0 -0.338262  1.119593  1.044367 -0.222187  0.499361 -0.246761  0.651583   

         V8        V9  ...       V21       V22       V23       V24       V25  \
0  0.098698  0.363787  ... -0.018307  0.277838 -0.110474  0.066928  0.128539   
1  0.085102 -0.255425  ... -0.225775 -0.638672  0.101288 -0.339846  0.167170   
2  0.247676 -1.514654  ...  0.247998  0.771679  0.909412 -0.689281 -0.327642   
3  0.377436 -1.387024  ... -0.108300  0.005274 -0.190321 -1.175575  0.647376   
4 -0.270533  0.817739  ... -0.009431  0.798278 -0.137458  0.141267 -0.206010   
5  0.260314 -0.568671  ... -0.208254 -0.559825 -0.026398 -0.371427 -0.232794   
6  0.081213  0.464960  ... -0.167716 -0.270710 -0.154104 -0.780055  0.750137   
7 -3.807864  0.615375  ...  1.943465 -1.015455  0.057504 -0.649709 -0.415267   
8  0.851084 -0.392048  ... -0.073425 -0.268092 -0.204233  1.011592  0.373205   
9  0.069539 -0.736727  ... -0.246914 -0.633753 -0.120794 -0.385050 -0.069733   

        V26       V27       V28  Amount  Class  
0 -0.189115  0.133558 -0.021053  149.62      0  
1  0.125895 -0.008983  0.014724    2.69      0  
2 -0.139097 -0.055353 -0.059752  378.66      0  
3 -0.221929  0.062723  0.061458  123.50      0  
4  0.502292  0.219422  0.215153   69.99      0  
5  0.105915  0.253844  0.081080    3.67      0  
6 -0.257237  0.034507  0.005168    4.99      0  
7 -0.051634 -1.206921 -1.085339   40.80      0  
8 -0.384157  0.011747  0.142404   93.20      0  
9  0.094199  0.246219  0.083076    3.68      0  

[10 rows x 31 columns]

6. Print last few rows

In [11]:
df.tail()

Time         V1         V2        V3        V4        V5  \
284802  172786.0 -11.881118  10.071785 -9.834783 -2.066656 -5.364473   
284803  172787.0  -0.732789  -0.055080  2.035030 -0.738589  0.868229   
284804  172788.0   1.919565  -0.301254 -3.249640 -0.557828  2.630515   
284805  172788.0  -0.240440   0.530483  0.702510  0.689799 -0.377961   
284806  172792.0  -0.533413  -0.189733  0.703337 -0.506271 -0.012546   

              V6        V7        V8        V9  ...       V21       V22  \
284802 -2.606837 -4.918215  7.305334  1.914428  ...  0.213454  0.111864   
284803  1.058415  0.024330  0.294869  0.584800  ...  0.214205  0.924384   
284804  3.031260 -0.296827  0.708417  0.432454  ...  0.232045  0.578229   
284805  0.623708 -0.686180  0.679145  0.392087  ...  0.265245  0.800049   
284806 -0.649617  1.577006 -0.414650  0.486180  ...  0.261057  0.643078   

             V23       V24       V25       V26       V27       V28  Amount  \
284802  1.014480 -0.509348  1.436807  0.250034  0.943651  0.823731    0.77   
284803  0.012463 -1.016226 -0.606624 -0.395255  0.068472 -0.053527   24.79   
284804 -0.037501  0.640134  0.265745 -0.087371  0.004455 -0.026561   67.88   
284805 -0.163298  0.123205 -0.569159  0.546668  0.108821  0.104533   10.00   
284806  0.376777  0.008797 -0.473649 -0.818267 -0.002415  0.013649  217.00   

        Class  
284802      0  
284803      0  
284804      0  
284805      0  
284806      0  

[5 rows x 31 columns]

7. Print columns of the dataframe

In [12]:
df.columns

Index(['Time', 'V1', 'V2', 'V3', 'V4', 'V5', 'V6', 'V7', 'V8', 'V9', 'V10',
       'V11', 'V12', 'V13', 'V14', 'V15', 'V16', 'V17', 'V18', 'V19', 'V20',
       'V21', 'V22', 'V23', 'V24', 'V25', 'V26', 'V27', 'V28', 'Amount',
       'Class'],
      dtype='object')

## Further Exploration

1. View the top n rows of the specific column

In [13]:
df[['Class', 'Time', 'Amount']].head()

Class  Time  Amount
0      0   0.0  149.62
1      0   0.0    2.69
2      0   1.0  378.66
3      0   1.0  123.50
4      0   2.0   69.99

2. View last few rows of Dataframe

In [14]:
df[['Class', 'Time', 'Amount']].tail()

Class      Time  Amount
284802      0  172786.0    0.77
284803      0  172787.0   24.79
284804      0  172788.0   67.88
284805      0  172788.0   10.00
284806      0  172792.0  217.00

3. Check the unique values of the class. 

The `Class` column, which decides if a transaction is fraudulent or not, should contain only two possible values (0,1), as we were informed by the client. It is useful to check the unique values of the class. 

Count the number of unique values using the .nunique() method on the Series (column) Class. 


In [15]:
df['Class'].nunique()

2

4. Check what these unique values are

After we know there are two unique values for the Class column, we can check what these unique values are:

[0 1] that shows there are two unique values for the Class column that are 0 and 1. 

In [16]:
df['Class'].unique()

array([0, 1], dtype=int64)

5. Find and store value counts

Also, we can use the `.value_counts()` method on the `Class` Series to know the value of Class and how often they occur.  We will perform this operation in this step and store the value counts in the `class_counts` variable.

Store the value counts in the variable defined as `id_counts` and then display the stored values

In [17]:
class_counts = df['Class'].value_counts() 
print(class_counts)

Class
0    284315
1       492
Name: count, dtype: int64


Based on the previous output, it shows all the class values have been reported by the client and there is no other value that is not defined by the data dictionary. 

6. Check rows that have missing values for specific features.

There are several ways to check missing values. Although it is not within the context of this subject to dive into details of how to deal with missing values, we just later mention an example of missing values and a possible way to explore and check them. 

We can create a class mask

In [18]:
class_mask = df['Class'].isnull()
print(class_mask[0:5])

0    False
1    False
2    False
3    False
4    False
Name: Class, dtype: bool


Output shows that `isnull()` method returns `False` if the value is missing for a specific row for a specific feature (`Class` in this case). Similarly, `isnull()` method returns `True` if the value is missing for a specific row for a specific feature.

You can also count the number of missing values

In [19]:
print(sum(class_mask))

0


- `notnull()` returns `True` : the value is not missing for a specific row for a specific feature (`Class` in this case)
- `notnull()` returns `False` : the value is missing for a specific row for a specific feature,

In [20]:
class_mask_notNull = df['Class'].notnull()
print(class_mask_notNull[0:5])
print(sum(class_mask_notNull))

0    True
1    True
2    True
3    True
4    True
Name: Class, dtype: bool
284807


7. Generate a tabular report of summary statistics of specific features

Outpur shows 
- count of Amount feature values
- mean
- standard deviation std
- minimum value min  
- maximum value max 
 
We can notice that there are `284807` values. The minimum amount is `0.0` and the maximum is `25691.16`. 

Analysing these values could provide good indicators on the feature itself and if it could help to achieve the goal of the business problem (e.g. predicting if credit transaction is fraudulent based on the amount spent). 

In [21]:
df[['Amount']].describe()

Amount
count  284807.000000
mean       88.349619
std       250.120109
min         0.000000
25%         5.600000
50%        22.000000
75%        77.165000
max     25691.160000